modules/seo/title-seo.php

```diff
@@ -0,0 +1,249 @@
+ *
+ * @return array Array of allowed token strings.
+ */
+function get_allowed_tokens() {
+	return array(
+		'front_page' => [ 'site_name', 'tagline' ],
+		'posts'      => [ 'site_name', 'tagline', 'post_title' ],
+		'pages'      => [ 'site_name', 'tagline', 'page_title' ],
+		'groups'     => [ 'site_name', 'tagline', 'group_title' ],
+		'archives'   => [ 'site_name', 'tagline', 'date' ],
+	);
+}
+
+/**
+ * Used to modify the default title with custom SEO title.
+ *
+ * @param string $default_title Default title for current page.
+ *
+ * @return string Custom title with replaced tokens or default title.
+ */
+function get_custom_title( $default_title = '' ) {
+	// Don't filter title for unsupported themes
+	if ( is_conflicted_theme() ) {
+		return $default_title;
+	}
+
+	$page_type = get_page_type();
+
+	// Keep default title if invalid page type is supplied
+	if ( empty( $page_type ) ) {
+		return $default_title;
+	}
+
+	$title_formats = get_option( 'advanced_seo_title_formats' );
+
+	// Keep default title if user has not defined custom title for this page type
+	if ( empty( $title_formats[ $page_type ] ) ) {
+		return $default_title;
+	}
+
+	if ( ! is_enabled_advanced_seo() ) {
+		return $default_title;
+	}
+
+	$custom_title = array_reduce( $title_formats[ $page_type ], function ( $title, $item ) {
+		return $title .= ( 'token' === $item['type'] )
+			? get_token_value( $item['value'] )
+			: $item['value'];
+	}, '' );
+
+	return esc_html( $custom_title );
+}
+
+/**
+ * Returns string value for given token.
+ *
+ * @param string $token_name The token name value that should be replaced.
+ *
+ * @return string Token replacement for current site, or empty string for unknown token name.
+ */
+function get_token_value( $token_name ) {
+
+	switch ( $token_name ) {
+		case 'site_name':
+			return get_bloginfo( 'name' );
+
+		case 'tagline':
+			return get_bloginfo( 'description' );
+
+		case 'post_title':
+		case 'page_title':
+			return get_the_title();
+
+		case 'group_title':
+			return single_tag_title( '', false );
+
+		case 'date':
+			return trim( single_month_title( ' ', false ) );
+
+		default:
+			return '';
+	}
+}
+
+/**
+ * Returns page type for current page. We need this helper in order to determine what
+ * user defined title format should be used for custom title.
+ *
+ * @return string|bool Type of current page or false if unsupported.
+ */
+function get_page_type() {
+
+	if ( is_front_page() ) {
+		return 'front_page';
+	}
+
+	if ( is_category() or is_tag() ) {
+		return 'groups';
+	}
+
+	if ( is_archive() ) {
+		return 'archives';
+	}
+
+	if ( is_page() ) {
+		return 'pages';
+	}
+
+	if ( is_singular() ) {
+		return 'posts';
+	}
+
+	return false;
+}
+
+/**
+ * Checks if current theme is defining custom title that won't work nicely
+ * with our custom SEO title override.
+ *
+ * @return bool True if current theme sets custom title, false otherwise.
+ */
+function is_conflicted_theme() {
+	$conflicted_themes = array(
```

Comment:
I was planning to remove this altogether since the theme names are wpcom specific, but adding a filter might be a better option.